Src/Common/WebHeaderCollectionExtensions.cs

```diff
@@ -124,6 +126,67 @@ public static IEnumerable<string> GetHeaderValue(this NameValueCollection header
             return null;
         }
 
+        public static void ReadActivityBaggage(this NameValueCollection headers, Activity activity)
+        {
+            Debug.Assert(headers != null, "Headers must not be null");
+            Debug.Assert(activity != null, "Activity must not be null");
+
+            int itemsCount = 0;
+            var correlationContexts = headers.GetValues(RequestResponseHeaders.CorrelationContextHeader);
+            if (correlationContexts == null || correlationContexts.Length == 0)
+            {
+                return;
+            }
+
+            int overallLength = 0;
+            foreach (var cc in correlationContexts)
+            {
+                var headerValue = cc.AsSpan();
+                int currentLength = 0;
+                int initialLength = headerValue.Length;
+                while (itemsCount < CorrelationContextMaxPairs && currentLength < initialLength)
+                {
+                    var nextSegment = headerValue.Slice(currentLength);
+                    var nextComma = nextSegment.IndexOf(',');
+                    if (nextComma < 0)
+                    {
+                        // last one
+                        nextComma = nextSegment.Length;
+                    }
+
+                    if (nextComma == 0)
+                    {
+                        currentLength += 1;
+                        overallLength += 1;
+                        continue;
+                    }
+
+                    if (overallLength + nextComma >= CorrelationContextHeaderMaxLength)
+                    {
+                        return;
+                    }
+
+                    ReadOnlySpan<char> kvp = nextSegment.Slice(0, nextComma).Trim();
+
+                    var separatorInd = kvp.IndexOf('=');
+                    if (separatorInd > 0 && separatorInd < kvp.Length - 1)
+                    {
+                        var separatorIndNext = kvp.Slice(separatorInd + 1).IndexOf('=');
+                        // check there is just one '=' in key-value-pair
+                        if (separatorIndNext < 0)
+                        {
+                            activity.AddBaggage(kvp.Slice(0, separatorInd).Trim().ToString(),
```

Comment:
nit: this line is difficult to read because of all the nested parenthesis. Consider changing this to use named parameters.